File: GFG/040125/Java/CountTriplet.java

// Problem Link: https://www.geeksforgeeks.org/problems/count-all-triplets-with-given-sum-in-sorted-array/1

class Solution {
    public int countTriplets(int[] arr, int target) {
        // Code Here
        int n= arr.length;
        int res=0;
        
        for(int i=0; i<n-2; i++){
            
            int left = i+1, right = n-1;
            
            while(left < right){
                
                int sum = arr[left]+arr[i]+arr[right];
                
                if(sum == target){
                    res++;
                    int start = left+1;
                    int end = right -1;
                    
                    while(start < right && arr[start] == arr[start-1]){
                        res++;
                        start++;
                    }
                    
                    while(left< end && arr[end] == arr[end+1]){
                        res++;
                        end--;
                    }
                    left++;
                    right--;
                }
                
                else if(sum > target) right-=1;
                else left+=1;
            }
        }
        
        return res;
    }
}

File: GFG/040125/Python/CountTriplet.py
#problem_link: https://www.geeksforgeeks.org/problems/count-all-triplets-with-given-sum-in-sorted-array/1

class Solution:
    def countTriplets(self, arr, target):
        # code here
        res=0
        n= len(arr)
        
        for i in range(len(arr)-2):
            left = i+1
            right = len(arr)-1
            
            while(left < right):
               sum = arr[i] + arr[left] + arr[right]
               
               if(sum == target):
                   res = res+1
               # count freuency
                 
                   start = left+1
                   end = right-1
                  
               
                   while start < right and arr[start] == arr[start-1]:
                         res+=1
                         start = start+1
                   while left < end and arr[end] == arr[end+1]:
                         end = end-1
                         res+=1
                   left+=1
                   right-=1
                         
                       
               elif(sum >target):
                    right= right-1
               else:
                 left = left+1
                    
                  
        
        return res
